tensorboard/backend/application.py

```diff
@@ -405,8 +418,11 @@ def _get_connect_params(query):
   return {k: json.loads(v[0]) for k, v in params.items()}
 
 
-def _clean_path(path):
-  """Removes trailing slash if present, unless it's the root path."""
-  if len(path) > 1 and path.endswith('/'):
+def _clean_path(path, _base_url=""):
+  """
+  If base_url is present, do not remove the ending '/'
+  If the base_url contains a '/' return as is, if not append '/'
+  """
+  if len(path) > 1 and path.endswith('/') and path != _base_url + '/':
```

Comment:
I'm a bit confused here. Why do we strip the ending '/' only if the path is not `base_url + '/'`?

The documentation seems to imply that `If base_url is present, do not remove the ending '/'`, but the code does not remove the final `/` if there is a route beyond that base, ie `/base_url/foo/`.

How does this logic sound?

```python
def _clean_path(path, path_prefix):
  """Cleans the path of the request.

  Removes the ending '/' if the request begins with the path prefix and
  pings a non-empty route.
  
  Arguments:
    path: The path of a request.
    path_prefix: The prefix string that every route of this TensorBoard instance starts with.

  Returns:
    The route to use to serve the request (with the path prefix stripped if applicable).
  """
  regex = re.compile(path_prefix + ".+/$")
  if regex.match(path):
    # Remove the ending prefix.
    return path[-1]
  return path
```

Reply:
> Why do we strip the ending '/' only if the path is not base_url + '/'?

Because the core plugin handles route `/`, which will become route `/foo/` with a path prefix of `/foo`. It's quite hacky. I think I'd prefer to just remove all trailing slashes everywhere internally (i.e., in the `self.data_applications` keyset). Then `_clean_path` is just `lambda path: path.rstrip('/')`. 

With the current semantics, using regex here is certainly overkill. Note that you'd need to actually write the pattern as `re.escape(path_prefix) + r'…'`. But there's no reason to use a regex engine at all when all you need are some simple string functions.

With the current semantics, I favor the implementation written in this PR, with the minor request to change `len(path) > 1` to `path != '/'`, so that the intent is clearer.

Reply:
Yes, that is more logical, the original code was an error.

Reply:
Yeah, William's formulation SG. `path != '/'` makes it clear.